prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/CustomFHIRFunctions.kt

```diff
@@ -319,4 +356,95 @@ object CustomFHIRFunctions {
         }
         return if (type != null) mutableListOf(StringType(type)) else mutableListOf()
     }
+
+    /**
+     * Get the LOINC Code from the LIVD table based on the device id, equipment model id, test kit name id, or the
+     * element model name
+     * @return a list with one value denoting the LOINC Code, or an empty list
+     */
+    fun livdTableLookup(
+        focus: MutableList<Base>,
+        parameters: MutableList<MutableList<Base>>?,
+        metadata: Metadata = Metadata.getInstance()
+    ): MutableList<Base> {
+        val lookupTable = metadata.findLookupTable(name = LivdLookup.livdTableName)
+
+        if (focus.size != 1) {
+            throw SchemaException("Must call the livdTableLookup function on a single observation")
+        }
+
+        val observation = focus.first()
+        if (observation !is Observation) {
+            throw SchemaException("Must call the livdTableLookup function on an observation")
+        }
+
+        var result: String? = ""
+        // Maps to OBX 17 CWE.1 Which is coding[1].code
+        val coding = (observation as Observation?)?.code?.coding
+        if (!coding.isNullOrEmpty()) {
+            coding[0]?.code?.let {
+                result = LivdLookup.find(
+                    testPerformedCode = null,
+                    processingModeCode = null,
+                    deviceId = it,
+                    equipmentModelId = null,
+                    testKitNameId = null,
+                    equipmentModelName = null,
+                    tableRef = lookupTable,
+                    tableColumn = parameters!!.first().first().primitiveValue()
+                )
+            }
+        }
+
+        // Maps to OBX 18 which is mapped to Device.identifier
+        val equipmentModelId = observation.device.identifier.id
+        if (!result.isNullOrBlank() && !equipmentModelId.isNullOrEmpty()) {
```

Comment:
We are using it because, if nothing has been found from the previous search then we want to try to search through other data that might be in the bundle to see if it yields results.